src/rcc/rec.rs

```diff
@@ -0,0 +1,256 @@
+                            /// Modify a kernel clock for this
+                            /// peripheral. See RM0433 Section 8.5.8.
+                            ///
+                            /// It is possible to switch this clock
+                            /// dynamically without generating spurs or
+                            /// timing violations. However, the user must
+                            /// ensure that both clocks are running. See
+                            /// RM0433 Section 8.5.10
+                            pub fn kernel_clk_mux(self, sel: [< $p ClkSel >]) -> Self {
+                                // unsafe: Owned exclusive access to this bitfield
+                                interrupt::free(|_| {
+                                    let ccip = unsafe {
+                                        &(*RCC::ptr()).[< $ccip r >]
+                                    };
+                                    ccip.modify(|_, w| w.
+                                                [< $p:lower sel >]().variant(sel));
+                                });
+                                self
+                            }
+
+                            #[inline(always)]
+                            #[allow(unused)]
+                            /// Return the current kernel clock selection
+                            pub fn get_kernel_clk_mux(&self) ->
+                                variant_return_type!([< $p ClkSel >] $(, $Variant)*)
+                            {
+                                // unsafe: We only read from this bitfield
+                                let ccip = unsafe {
+                                    &(*RCC::ptr()).[< $ccip r >]
+                                };
+                                ccip.read().[< $p:lower sel >]().variant()
+                            }
+                        )*
+                    }
+
+                    $(
+                        #[doc=$clk_doc]
+                        pub type [< $p ClkSel >] =
+                            rcc::[< $ccip r >]::[< $p:upper SEL_A >];
+                    )*
+                )*
+            )+
+        }
+    }
+}
+
+// If the PAC does not fully specify a CCIP field (perhaps because one or
+// more values are reserved), then we use a different return type
+macro_rules! variant_return_type {
+    ($t:ty) => { $t };
+    ($t:ty, $Variant: ident) => {
+        stm32h7::Variant<u8, $t>
+    };
+}
+
+// Must only contain peripherals that are not used anywhere in this HAL
+peripheral_reset_and_enable_control! {
+    AHB1, "AMBA High-performance Bus (AHB1) peripherals" => [
+        Eth1Mac, Dma2, Dma1,
+        #[cfg(any(feature = "dualcore"))] Art
+    ];
+    AHB2, "AMBA High-performance Bus (AHB2) peripherals" => [
+        Sdmmc2, Hash, Crypt
+    ];
+    AHB3, "AMBA High-performance Bus (AHB3) peripherals" => [
+        Sdmmc1,
+        Qspi kernel d1ccip "QUADSPI kernel clock source selection",
```

Comment:
The syntax for the clock allowances on a peripheral don't make it clear, perhaps we could make this something like `Qspi: [kernel, d1ccip], description: "QUADSPI kernel clock source selection"` or something similar when the clocks are allowed to be configured? This may be restricted by the macro structure, but just thought I'd throw this in.

Reply:
It is quite a serious macro unfortunately. The macro will support more punctuation, but I'm lazy and don't want lines to wrap (80 characters). Perhaps adding a colon after kernel makes sense?